perspective calc() values aren't clipped to 0

graouts opened the pull request:
#### c1c0e7fbede03d23c5191147fbd385498a289301
<pre>
perspective calc() values aren&apos;t clipped to 0
<a href="https://bugs.webkit.org/show_bug.cgi?id=249151">https://bugs.webkit.org/show_bug.cgi?id=249151</a>

Reviewed by Simon Fraser and Dean Jackson.

The &quot;perspective&quot; property is specified to only allow non-negative values,
see <a href="https://www.w3.org/TR/css-transforms-2/#perspective-property.">https://www.w3.org/TR/css-transforms-2/#perspective-property.</a> However,
we don&apos;t clip values that are computed to less than 0, for instance using a
calc() expression. Instead, we deem them invalid internally and thus consider
as if &quot;none&quot; was provided, which has implications when blending. Indeed, when
interpolating between a valid value and &quot;none&quot;, we use discrete interpolation.

We now correctly clip values below 0.

Note that the relevant WPT test still fails, because the test itself is not
valid. It is being addressed in a dedicated WPT PR and will be imported back
into WebKit yielding a PASS result. See <a href="https://github.com/web-platform-tests/wpt/pull/37457.">https://github.com/web-platform-tests/wpt/pull/37457.</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/perspective-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/responsive/perspective-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumePerspective):

Canonical link: <a href="https://commits.webkit.org/257779@main">https://commits.webkit.org/257779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f66226d5bb59d2385e2c68b669d40848878430fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99997 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9166 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33077 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109344 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169577 "Failed to checkout and rebase branch from PR 7493") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103996 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10050 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86480 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92444 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107244 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105765 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/2961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2921 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/46155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/9055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/43259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4769 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2746 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -7497,7 +7497,8 @@
                 "custom-parser": true,
-                "parser-requires-context-mode": true
+                "parser-requires-context-mode": true,
+                "parser-grammar": ["<length [0,inf]>"]
```
Why does this have both `custom-parser` and `parser-grammar`? @weinig Is this even supported?

graouts replied:
I don’t actually know what this change does, but saw other places that did something similar. I know that the .cpp change alone suffices. I could just remove this.

nt1m replied:
`"parser-grammar": ["<length [0,inf]>"]` actually automatically generates code for parsing. So you probably don't want to use it along with `custom-parser: true`. @weinig, it might be worth adding a warning or build error if someone does this.

graouts replied:
I'll try removing the `custom-parser` bit and see what happens :)

graouts replied:
I removed the `CSSProperties.json` changes for now, they weren't material to the fix.

weinig replied:
The custom parser is still different from the grammar you specified as it supports "none". So if you want them to match it should be:

"parser-grammar": "none | <length [0,inf]>"

I think make it a failure to have both a grammar and custom-parsing makes sense. I'll do that.

weinig replied:
(Also, don't need the surrounding array: [...] in the grammar property)

weinig replied:
> I removed the `CSSProperties.json` changes for now, they weren't material to the fix.

I had the browser window open too long. Ignore my comments).

graouts committed:
perspective calc() values aren't clipped to 0
https://bugs.webkit.org/show_bug.cgi?id=249151

Reviewed by Simon Fraser and Dean Jackson.

The "perspective" property is specified to only allow non-negative values,
see https://www.w3.org/TR/css-transforms-2/#perspective-property. However,
we don't clip values that are computed to less than 0, for instance using a
calc() expression. Instead, we deem them invalid internally and thus consider
as if "none" was provided, which has implications when blending. Indeed, when
interpolating between a valid value and "none", we use discrete interpolation.

We now correctly clip values below 0.

Note that the relevant WPT test still fails, because the test itself is not
valid. It is being addressed in a dedicated WPT PR and will be imported back
into WebKit yielding a PASS result. See https://github.com/web-platform-tests/wpt/pull/37457.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/perspective-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/responsive/perspective-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumePerspective):

Canonical link: https://commits.webkit.org/257779@main
